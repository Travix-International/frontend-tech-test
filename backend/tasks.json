{
  "tasks": [
    {
      "id": 1,
      "title": "Buy Food",
      "description": "Language is a system that consists of the development, acquisition, maintenance and use of complex systems of communication, particularly the human ability to do so; a language is any specific example of such a system. The scientific study of language is called linguistics."
    },
    {
      "id": 2,
      "title": "Feed the Dog",
      "description": "The domestic dog is a member of the genus Canis, which forms part of the wolf-like canids, and is the most widely abundant terrestrial carnivore."
    },
    {
      "id": 3,
      "title": "Travel",
      "description": "Travel is the movement of people between distant geographical locations. Travel can be done by foot, bicycle, automobile, train, boat, bus, airplane, ship or other means, with or without luggage, and can be one way or round trip. Travel can also include relatively short stays between successive movements."
    },
    {
      "id": 4,
      "title": "Nature",
      "description": "Nature is a British multidisciplinary scientific journal, first published on 4 November 1869. It is one of the most recognizable scientific journals in the world, and was ranked the world's most cited."
    },
    {
      "id": 5,
      "title": "Buy Food",
      "description": "Language is a system that consists of the development, acquisition, maintenance and use of complex systems of communication, particularly the human ability to do so; a language is any specific example of such a system. The scientific study of language is called linguistics."
    },
    {
      "id": 6,
      "title": "Feed the Dog",
      "description": "The domestic dog is a member of the genus Canis, which forms part of the wolf-like canids, and is the most widely abundant terrestrial carnivore."
    },
    {
      "id": 7,
      "title": "Travel",
      "description": "Travel is the movement of people between distant geographical locations. Travel can be done by foot, bicycle, automobile, train, boat, bus, airplane, ship or other means, with or without luggage, and can be one way or round trip. Travel can also include relatively short stays between successive movements."
    },
    {
      "id": 8,
      "title": "Nature",
      "description": "Nature is a British multidisciplinary scientific journal, first published on 4 November 1869. It is one of the most recognizable scientific journals in the world, and was ranked the world's most cited."
    }
  ]
}
