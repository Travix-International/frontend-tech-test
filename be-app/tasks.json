{
  "tasks": [
    {
      "id": 1,
      "title": "Task",
      "description": "foo bar description"
    }
  ]
}