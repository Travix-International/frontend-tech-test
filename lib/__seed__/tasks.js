export const tasks = [
  {
    _id: '1',
    title: 'Wash the dishes',
    description: 'Yet another generic task',
    isComplete: false
  },
  {
    _id: '2',
    title: 'Do the laundry',
    description: 'Existence is pain',
    isComplete: true
  },
  {
    _id: '3',
    title: 'Take over the world',
    description: 'Send an invite to Gru and Megamind',
    isComplete: false
  }
];
