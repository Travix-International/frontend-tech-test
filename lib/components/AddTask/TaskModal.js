import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as taskActions from '../../actions/taskActions';
import styles from './TaskModal.scss';
import Modal from '../Modal/Modal';

class Task extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    addModal: PropTypes.bool.isRequired,
    editModal: PropTypes.string,
    tasks: PropTypes.array.isRequired
  };

  static defaultProps = {
    editModal: null
  };

  constructor(props) {
    super(props);
    this.titleRef = React.createRef();
    this.descriptionRef = React.createRef();
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { editModal, actions } = this.props;
    const title = this.titleRef.current.value;
    const description = this.descriptionRef.current.value;
    if (title.trim().length) {
      const task = { title, description };
      if (editModal !== null) {
        actions.updateTask(editModal, task);
      } else {
        actions.createTask(task);
      }
      this.handleModalClose();
    } else {
      this.titleRef.current.focus();
    }
  };

  handleModalClose = () => {
    this.setState({ title: '', description: '' });
    if (this.props.editModal) {
      this.props.actions.toggleEditModal();
    } else {
      this.props.actions.toggleAddModal();
    }
  };

  render() {
    const { editModal, addModal, tasks } = this.props;
    const isEdit = editModal !== null;
    let title = '';
    let description = '';
    const submitButtonText = isEdit ? 'Edit' : 'Add';

    if (isEdit) {
      const task = tasks.find(t => t._id === editModal);
      title = task.title;
      description = task.description;
    }

    return (
      <Modal
        onClose={this.handleModalClose}
        show={addModal || isEdit}
        styles={styles.root}
      >
        <form onSubmit={this.handleSubmit}>
          <input
            className={styles.title}
            defaultValue={title}
            id="todo-text"
            name="title"
            placeholder="title"
            ref={this.titleRef}
            type="text"
          />
          <textarea
            className={styles.description}
            defaultValue={description}
            id="todo-description"
            maxLength="140"
            name="description"
            placeholder="description (max 140 characters)"
            ref={this.descriptionRef}
            rows="5"
          />
          <input
            className={styles.button}
            type="submit"
            value={submitButtonText}
          />
        </form>
      </Modal>
    );
  }
}

const mapStateToProps = ({ addModal, editModal, tasks }) => ({
  addModal,
  editModal,
  tasks
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(taskActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Task);
