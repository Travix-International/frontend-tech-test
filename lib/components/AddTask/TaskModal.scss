@import '../../scss/_settings.scss';
@import '../../scss/_objects.scss';

.title,
.description {
  @extend .text-box;

  color: $color-light;
}

.title {
  font-size: 4rem;
  margin-bottom: $spacing-1 * 2;
}

.description {
  font-size: 2.5rem;
  resize: none;
}

.root {
  background: $color-brand;
}

.button {
  @extend .button;

  color: lighten($color-brand-light, 10);
  font-size: 4.5rem;
  margin-top: $spacing-1 * 2;
}
