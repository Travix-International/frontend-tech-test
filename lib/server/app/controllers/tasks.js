import pick from 'lodash.pick';
import emitter from 'central-event';

import Task from '../models/Task';

/**
 * Helper middleware for
 * PUT/PATCH/DELETE /api/tasks/:id
 *
 * id: ObjectId
 *
 * Gets the requested task by its ID from db and sets it on the request object.
 * Thus avoiding repeating the db request logic.
 *
 * If found return status code 200 and the resource.
 * If not found return status code 404.
 */
export const helperMiddleware = async (req, res, next) => {
  const task = await Task.findById(req.params.id);
  if (task) {
    req.task = task;
    next();
  } else {
    res.status(404).send({ message: 'Not found' });
  }
};

/**
 * GET /api/tasks
 *
 * Return the list of tasks with status code 200.
 */
export const fetch = async (req, res) => {
  try {
    const tasks = await Task.find({});
    res.status(200).json(tasks);
  } catch (error) {
    res.status(500).send({
      message: error.message
    });
  }
};

/**
 * POST /api/tasks
 *
 * title: string
 * description: string
 *
 * Add a new task to the DB with the given title and description.
 * Notify Websockets server to push new data to clients
 * Return status code 201.
 */
export const create = (req, res) => {
  const task = new Task(req.body);
  // TODO: Sanitize title, description
  if (!req.body.title) {
    res.status(400).send({ message: 'Task is required' });
  } else {
    task.save();
    emitter.emit('tasks/updated');
    res.status(201).send(task);
  }
};

/**
 * PATCH /api/tasks/:id
 *
 * id: ObjectId
 * title: string
 * description: string
 * isComplete: boolean
 *
 * Update the task with the given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and updated as well, return a status code 200.
 */
export const update = async (req, res) => {
  if (req.body._id) {
    delete req.body._id;
  }

  Object.keys(req.body).forEach((prop) => {
    req.task[prop] = req.body[prop];
  });

  try {
    await req.task.save();
    emitter.emit('tasks/updated');
    res.status(200).json(req.task);
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};

/**
 * PUT /api/tasks/:id
 *
 * id: ObjectId
 * title: string
 * description: string
 *
 * Update the task with the given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and updated as well, return a status code 200.
 */
export const replace = async (req, res) => {
  Object.assign(
    req.task,
    pick(req.body, ['text', 'description', 'isComplete'])
  );

  try {
    await req.task.save();
    emitter.emit('tasks/updated');
    res.status(200).json(req.task);
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};

/**
 * DELETE /task/:id
 *
 * id: ObjectId
 *
 * Delete the task linked to the  given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and deleted as well, return a status code 204.
 */
export const remove = async (req, res) => {
  try {
    await req.task.remove();
    emitter.emit('tasks/updated');
    res.status(204).send({ message: 'Removed' });
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};
