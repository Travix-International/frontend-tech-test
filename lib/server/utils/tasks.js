import axios from 'axios';

/* Helper to get tasks using GET API */
export const fetchTasks = async () => {
  const requestURL = `${process.env.HOST}:${process.env.PORT}`;
  const response = await axios.get(`${requestURL}/api/tasks`);
  return response.data;
};
