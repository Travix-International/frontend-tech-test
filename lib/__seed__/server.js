import { ObjectID } from 'mongodb';
import Task from '../server/app/models/Task';

import { tasks } from './tasks';

const dbTasks = tasks.map((task) => {
  task._id = new ObjectID();
  return task;
});

export const populateTasks = (done) => {
  /* eslint-disable promise/no-callback-in-promise, no-console */
  Task.remove({})
    .then(() => Task.insertMany(dbTasks))
    .then(() => done())
    .catch(error => console.error(error));
};
