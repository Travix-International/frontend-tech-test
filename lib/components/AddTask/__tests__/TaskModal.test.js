import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { TaskModal } from '../TaskModal';
import * as tasksSeed from '../../../__seed__/tasks';

const testProps = {
  actions: {},
  addModal: true,
  editModal: null,
  tasks: tasksSeed.tasks
};

describe('TaskModal', () => {
  const mockUpdateTaskFn = jest.fn();
  const mockCreateTaskFn = jest.fn();
  const mockToggleAddModalFn = jest.fn();
  const mockToggleEditModalFn = jest.fn();

  testProps.actions.updateTask = mockUpdateTaskFn;
  testProps.actions.createTask = mockCreateTaskFn;
  testProps.actions.toggleAddModal = mockToggleAddModalFn;
  testProps.actions.toggleEditModal = mockToggleEditModalFn;

  it('renders correctly', () => {
    const tree = renderer.create(<TaskModal {...testProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls create task function when Add Task form is submitted', () => {
    const wrapper = mount(<TaskModal {...testProps} />);
    const task = { title: 'Task 1', description: 'A brief description' };
    wrapper.find('#todo-text').instance().value = task.title;
    wrapper.find('#todo-description').instance().value = task.description;
    wrapper
      .find('form')
      .first()
      .simulate('submit');
    expect(mockCreateTaskFn).toBeCalled();
    expect(mockCreateTaskFn).toBeCalledWith(task);
  });

  it('calls update task function when Add Task form is submitted with an editModal _id passed in props', () => {
    const props = { ...testProps, editModal: testProps.tasks[0]._id };
    const wrapper = mount(<TaskModal {...props} />);
    const updatedTitle = 'Task 2';
    wrapper.find('#todo-text').instance().value = updatedTitle;
    wrapper
      .find('form')
      .first()
      .simulate('submit');
    expect(mockUpdateTaskFn).toBeCalled();
    expect(mockUpdateTaskFn).toBeCalledWith(testProps.tasks[0]._id, {
      description: testProps.tasks[0].description,
      title: updatedTitle
    });
  });
});
