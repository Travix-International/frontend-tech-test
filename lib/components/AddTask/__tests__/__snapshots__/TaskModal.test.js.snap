// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskModal renders correctly 1`] = `
<div
  className="root root"
>
  <div
    className="content"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="title"
        defaultValue=""
        id="todo-text"
        name="title"
        placeholder="title"
        type="text"
      />
      <textarea
        className="description"
        defaultValue=""
        id="todo-description"
        maxLength="140"
        name="description"
        placeholder="description (max 140 characters)"
        rows="5"
      />
      <input
        className="button"
        type="submit"
        value="Add"
      />
    </form>
  </div>
  <div
    className="footer"
  >
    <button
      className="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</div>
`;
